imap_processing/swapi/l1/swapi_l1.py

```diff
@@ -1 +1,565 @@
+
+
+def get_indices_of_full_sweep(seq_number: np.ndarray):
+    """Get indices of full cycles.
+
+    Beginning of a sweep is marked by SWP_SCI.SEQ_NUMBER=0
+    (Sequence number of set of steps in energy sweep);
+    end of a sweep is marked by SWP_SCI.SEQ_NUMBER=11;
+    all packets must be present to process a sweep.
+
+    In this function, we get the indices of SEQ_NUMBER
+    0 and then construct full sweep indices.
+
+    Parameters
+    ----------
+    seq_number : numpy.ndarray
+        Array that contains SEQ_NUMBER data informations.
+        Eg. sci_dataset["SEQ_NUMBER"].data
+
+    Returns
+    -------
+    numpy.ndarray
+        1D array with indices of full cycle data.
+    """
+    indices_of_start = find_sweep_starts(seq_number)
+    # find_sweep_starts[..., None] creates array of shape(n, 1).
+    #   Eg. [[3], [8]]
+    # np.arange(12)[None, ...] creates array of shape(1, 12)
+    #   Eg. [[0, 1, 2, 3, ....., 11]]
+    # then we add both of them together to get an array of shape(n, 4)
+    #   Eg. [[3, 4, 5, 6,...14], [8, 9, 10, 11, ..., 19]]
+    full_cycles_indices = indices_of_start[..., None] + np.arange(12)[None, ...]
+    return full_cycles_indices.reshape(-1)
+
+
+def filter_full_cycle_data(full_cycle_data_indices: np.ndarray, l1a_data: xr.Dataset):
+    """Filter metadata and science of packets that makes full cycles.
+
+    Parameters
+    ----------
+    full_cycle_data_indices : numpy.ndarray
+        Array with indices of full cycles.
+    l1a_data : xarray.dataset
+        L1A dataset
+
+    Returns
+    -------
+    xarray.dataset
+        L1A dataset with filtered metadata.
+    """
+    # Had to create new xr.Dataset because Epoch shape and new data variables shapes was
+    # different.
+    full_sweep_dataset = xr.Dataset(
+        coords={"Epoch": l1a_data["Epoch"].data[full_cycle_data_indices]}
+    )
+    for key, value in l1a_data.items():
+        full_sweep_dataset[key] = xr.DataArray(value.data[full_cycle_data_indices])
+    return full_sweep_dataset
+
+
+def uncompress_count(count_data: np.ndarray, compress_indicator: np.ndarray = None):
+    """Uncompress counts based on compression indicators.
+
+    Parameters
+    ----------
+    count_data : numpy.ndarray
+        Array with counts.
+    compress_indicator : numpy.ndarray
+        Array with compression indicators.
+
+    Returns
+    -------
+    numpy.ndarray
+        Array with uncompressed counts.
+    """
+    # Uncompress counts based on compression indicators
+    # If 0, value is already uncompressed. If 1, value is compressed.
+    # If 1 and count is 0xFFFF, value is overflow.
+    new_count = copy.deepcopy(count_data)
+    compressed_count_indices = np.where(compress_indicator == 1)[0]
+
+    for index in compressed_count_indices:
+        if count_data[index] == 0xFFFF:  # Overflow
+            new_count[index] = -1
+        elif count_data[index] != 0xFFFF:
+            new_count[index] = count_data[index] * 16
+    return new_count
```

Comment:
I think you can avoid the for loop with logical ands for your conditions. You could store the intermediaries and invert the logic of these too depending on what you want to do.
```suggestion
    new_count[compress_indicator.astype(bool) & (count_data == 0xFFFF)] = -1
    new_count[compress_indicator.astype(bool) & (count_data != 0xFFFF)] *= 16

    return new_count
```

Reply:
I see what Greg is suggesting here, but it could be benficial to add error statements to the `if` and `elif` if you need to debug.